pkg/genericapiserver/openapi/openapi.go

```diff
@@ -0,0 +1,485 @@
+		for _, param := range route.ParameterDocs {
+			if routeParamMap[param.Data().Name] {
+				return commonParamsMap, fmt.Errorf("Duplicate parameter %v for route %v.", param.Data().Name, route)
+			}
+			paramOpsCountMap[param.Data().Name]++
+			paramDataMap[param.Data().Name] = param.Data()
+		}
+	}
+	for paramName, count := range paramOpsCountMap {
+		if count == len(routes) {
+			newParam, err := buildParameter(paramDataMap[paramName])
+			if err != nil {
+				return commonParamsMap, err
+			}
+			commonParamsMap[paramName] = newParam
+		}
+	}
+	return commonParamsMap, nil
+}
+
+func buildParameter(restParam restful.ParameterData) (ret spec.Parameter, err error) {
+	ret = spec.Parameter{
+		ParamProps: spec.ParamProps{
+			Name:        restParam.Name,
+			Description: restParam.Description,
+			Required:    restParam.Required,
+		},
+	}
+	switch restParam.Kind {
+	case restful.BodyParameterKind:
+		ret.In = "body"
+		ret.Schema = &spec.Schema{
+			SchemaProps: spec.SchemaProps{
+				Ref: spec.MustCreateRef("#/definitions/" + restParam.DataType),
+			},
+		}
+		return ret, nil
+	case restful.PathParameterKind:
+		ret.In = "path"
+		if !restParam.Required {
+			return ret, fmt.Errorf("Path parameters should be marked at required for parameter %v", restParam)
+		}
+	case restful.QueryParameterKind:
+		ret.In = "query"
+	case restful.HeaderParameterKind:
+		ret.In = "header"
+	case restful.FormParameterKind:
+		ret.In = "form"
+	default:
+		return ret, fmt.Errorf("Unknown restful operation kind : %v", restParam.Kind)
+	}
+	if !isSimpleDataType(restParam.DataType) {
+		return ret, fmt.Errorf("Restful DataType should be a simple type, but got : %v", restParam.DataType)
+	}
+	ret.Type = restParam.DataType
+	ret.Format = restParam.DataFormat
+	ret.UniqueItems = !restParam.AllowMultiple
+	// TODO(mbohlool): make sure they type of default value matches Type
+	if restParam.DefaultValue != "" {
+		ret.Default = restParam.DefaultValue
+	}
+
+	return ret, nil
+}
+
+func buildParameters(restParam []*restful.Parameter) (ret []spec.Parameter, err error) {
+	ret = make([]spec.Parameter, len(restParam))
+	for i, v := range restParam {
+		ret[i], err = buildParameter(v.Data())
+		if err != nil {
+			return ret, err
+		}
+	}
+	return ret, nil
+}
+
+func isSimpleDataType(typeName string) bool {
+	switch typeName {
+	// Note that "file" intentionally kept out of this list as it is not being used.
+	// "file" type has more requirements.
+	case "string", "number", "integer", "boolean", "array":
+		return true
+	}
+	return false
+}
+
+func getFloat64OrNil(str string) (*float64, error) {
+	if len(str) > 0 {
+		num, err := strconv.ParseFloat(str, 64)
+		return &num, err
+	}
+	return nil, nil
+}
+
+// TODO(mbohlool): Convert default value type to the type of parameter
+func getDefaultValue(str swagger.Special) interface{} {
+	if len(str) > 0 {
+		return str
+	}
+	return nil
+}
+
+func buildOpenAPIType(oldType *string, oldFormat string) ([]string, string, error) {
```

Comment:
Why name this buildOpenAPIType, instead of just buildType like others (buildModel, buildProperty). 
Though, I personally liked this name better, am fine either way - just be consistent.